ft-sdk/src/auth/session.rs

```diff
@@ -4,56 +4,272 @@ pub struct SessionID(pub String);
+/// * There is an issue querying the database.
+/// * There is an error creating the `http::HeaderValue`.
+#[cfg(feature = "auth-provider")]
+pub fn set_session_cookie(
+    conn: &mut ft_sdk::Connection,
+    ft_sdk::auth::SessionID(session_id): ft_sdk::auth::SessionID,
+    host: ft_sdk::Host,
+) -> Result<http::HeaderValue, ft_sdk::Error> {
+    use diesel::prelude::*;
+    use ft_sdk::auth::fastn_session;
+
+    let now = ft_sdk::env::now();
+
+    // Query to check if the session exists and get its expiration time.
+    let max_age = match fastn_session::table
+        .select(fastn_session::expires_at.nullable())
+        .filter(fastn_session::id.eq(&session_id))
+        .first::<Option<chrono::DateTime<chrono::Utc>>>(conn)
+    {
+        // If the session has an expiration time and it is in the future.
+        Ok(Some(session_expires_at)) if session_expires_at > now => {
+            let duration = session_expires_at - now;
+            cookie::time::Duration::new(duration.num_seconds(), duration.subsec_nanos())
+        }
+        // If the session does not have an expiration time.
+        Ok(None) => cookie::time::Duration::seconds(34560000),
+        // If the session has an expiration time and it is in the past.
+        Ok(_) => return Err(SessionIDError::SessionExpired(session_id.clone()).into()),
+        // If the session is not found.
+        Err(diesel::NotFound) => return Err(SessionIDError::SessionNotFound.into()),
+        // If there is an error querying the database.
+        Err(e) => return Err(e.into()),
+    };
+
+    // Build the cookie with the determined max age
+    let cookie = cookie::Cookie::build((ft_sdk::auth::SESSION_KEY, session_id))
+        .domain(host.without_port())
+        .path("/")
+        .max_age(max_age)
+        .same_site(cookie::SameSite::Strict)
+        .build();
+
+    // Convert the cookie to an HTTP header value and return it
+    Ok(http::HeaderValue::from_str(cookie.to_string().as_str())?)
+}
+
+/// Expires the session cookie immediately by setting its expiration time to the current time.
+#[cfg(feature = "auth-provider")]
+pub fn expire_session_cookie(host: ft_sdk::Host) -> Result<http::HeaderValue, ft_sdk::Error> {
+    let cookie = cookie::Cookie::build((ft_sdk::auth::SESSION_KEY, ""))
+        .domain(host.without_port())
+        .path("/")
+        .expires(convert_now_to_offsetdatetime())
+        .build();
+
+    Ok(http::HeaderValue::from_str(cookie.to_string().as_str())?)
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SessionIDError {
+    #[error("session not found")]
+    SessionNotFound,
+    #[error("session expired")]
+    SessionExpired(String),
+    #[error("failed to query db: {0:?}")]
+    DatabaseError(#[from] diesel::result::Error),
+}
+
+impl SessionID {
+    /// Retrieves a session ID based on a given user ID.
+    ///
+    /// This method queries the database to find a session associated with the
+    /// given user ID.
+    pub fn from_user_id(
```

Comment:
This is not from user_id, this if for user id. You create a new session for a user, not create a session from user. How is this different from `login()`?

Reply:
It fetches the session_id for the given user_id from the `fastn_session ` table.

Reply:
A user can have many sessions. This is wrong, it should return a list of session ids. Or it should be called `get_all_sessions_of_user` or `get_first/last_session_of_user`.

Reply:
Then why do we have [`SetUserIDError::MultipleSessionsFound`](https://github.com/FifthTry/ft-sdk/blob/main/ft-sdk/src/auth/session.rs#L30) error?

Reply:
That is doing the right thing, it is updating session table using session id. Since session id is unique this error should never be thrown, but we have added it instead of panicking to handle that case where diesel returns modified count > 1.